src/backend/parser/parse_coerce.c

```diff
@@ -1437,8 +1439,9 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 				pcategory = ncategory;
 				pispreferred = nispreferred;
 			}
-			else if (ncategory != pcategory
-				&& sql_dialect != SQL_DIALECT_TSQL) /* T-SQL allows to select common datatype between different categories */
+			else if (ncategory != pcategory &&
+				sql_dialect != SQL_DIALECT_TSQL && /* T-SQL allows to select common datatype between different categories */
+				(!dump_restore || (dump_restore && strcmp(dump_restore, "on") != 0))) /* allow common datatype between different categories if dump_restore GUC is set */
```

Comment:
comment should say "when restoring TSQL database", also dump_restore alone does not carry TSQL context, so we should alway specify that explicitly

Reply:
Yup, updated the comment!